src/sim.cpp

```diff
@@ -386,49 +388,66 @@ inline void collectObservationsSystem(Engine &ctx,
 
     Quat to_view = rot.inv();
 
-#pragma unroll
-    for (CountT i = 0; i < consts::numAgents - 1; i++) {
-        Entity other = other_agents.e[i];
-
-        Vector3 other_pos = ctx.get<Position>(other);
-        GrabState other_grab = ctx.get<GrabState>(other);
-        Vector3 to_other = other_pos - pos;
-
-        partner_obs.obs[i] = {
-            .polar = xyToPolar(to_view.rotateVec(to_other)),
-            .isGrabbing = other_grab.constraintEntity != Entity::none() ?
-                1.f : 0.f,
-        };
+    // Context::iterateQuery() runs serially over archteypes
+    // matching the components on which it is templated.
+    {
+        int idx = 0; // Context::iterateQuery() is serial, so this is safe.
+		ctx.iterateQuery(ctx.data().otherAgentQuery,
+			[&](Position& other_pos, GrabState& other_grab) {
+				Vector3 to_other = other_pos - pos;
+
+				// Detect whether or not the other agent is the current agent.
+				if (to_other.length() == 0.0f) {
+					return;
+				}
+
+				partner_obs.obs[idx++] = {
+					.polar = xyToPolar(to_view.rotateVec(to_other)),
+					.isGrabbing = other_grab.constraintEntity != Entity::none() ? 1.f : 0.f,
+				};
+			});
     }
 
-    const LevelState &level = ctx.singleton<LevelState>();
-    const Room &room = level.rooms[cur_room_idx];
-
-    for (CountT i = 0; i < consts::maxEntitiesPerRoom; i++) {
-        Entity entity = room.entities[i];
-
-        EntityObservation ob;
-        if (entity == Entity::none()) {
-            ob.polar = { 0.f, 1.f };
-            ob.encodedType = encodeType(EntityType::None);
-        } else {
-            Vector3 entity_pos = ctx.get<Position>(entity);
-            EntityType entity_type = ctx.get<EntityType>(entity);
-
-            Vector3 to_entity = entity_pos - pos;
-            ob.polar = xyToPolar(to_view.rotateVec(to_entity));
-            ob.encodedType = encodeType(entity_type);
-        }
-
+    // Previously, we iterated over entities that were explicitly members of the current room.
```

Comment:
This repo is intended to be a reference for using Madrona to a new user. The explanation here is fine but don't refer to "previously" just explain what the code is doing now if necessary.